src/c++/perf_analyzer/test_concurrency_manager.cc

```diff
@@ -0,0 +1,306 @@
+// (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+// OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+
+#include "command_line_parser.h"
+#include "concurrency_manager.h"
+#include "doctest.h"
+#include "mock_client_backend.h"
+#include "mock_model_parser.h"
+#include "test_load_manager_base.h"
+
+namespace triton { namespace perfanalyzer {
+
+class TestConcurrencyManager : public TestLoadManagerBase,
+                               public ConcurrencyManager {
+ public:
+  TestConcurrencyManager(
+      PerfAnalyzerParameters params, bool is_sequence_model = false,
+      bool use_mock_infer = false)
+      : use_mock_infer_(use_mock_infer),
+        TestLoadManagerBase(params, is_sequence_model),
+        ConcurrencyManager(
+            params.async, params.streaming, params.batch_size,
+            params.max_threads, params.max_concurrency, params.sequence_length,
+            params.shared_memory_type, params.output_shm_size,
+            params.start_sequence_id, params.sequence_id_range,
+            params.string_length, params.string_data, params.zero_input,
+            params.user_data, GetParser(), GetFactory())
+  {
+  }
+
+  void Infer(
+      std::shared_ptr<ThreadStat> thread_stat,
+      std::shared_ptr<ThreadConfig> thread_config) override
+  {
+    if (use_mock_infer_) {
+      return MockInfer(thread_stat, thread_config);
+    } else {
+      return ConcurrencyManager::Infer(thread_stat, thread_config);
+    }
+  }
+
+  // Mock out most of the complicated Infer code
+  //
+  void MockInfer(
+      std::shared_ptr<ThreadStat> thread_stat,
+      std::shared_ptr<ThreadConfig> thread_config)
+  {
+    if (!execute_) {
+      thread_config->is_paused_ = true;
+    }
+  }
+
+  /// Test that the correct Infer function is called in the backend
+  ///
+  void TestInferType()
+  {
+    // FIXME TMA-982: This delay is to avoid deadlock. Investigate why delay is
+    // needed.
+    stats_->request_delay = std::chrono::milliseconds(50);
+
+    ChangeConcurrencyLevel(params_.max_concurrency);
+
+    std::this_thread::sleep_for(std::chrono::milliseconds(500));
+
+    CheckInferType();
+  }
+
+  /// Test that the correct concurrency is maintained in the load manager
+  ///
+  void TestConcurrency(
+      std::chrono::milliseconds request_delay,
+      std::chrono::milliseconds sleep_time)
+  {
+    stats_->request_delay = request_delay;
+
+    ChangeConcurrencyLevel(params_.max_concurrency);
+    std::this_thread::sleep_for(sleep_time);
+
+    CheckConcurrency();
+  }
+
+  /// Test sequence handling
+  ///
+  void TestSequences()
+  {
+    auto sleep_time = std::chrono::milliseconds(500);
+
+    ChangeConcurrencyLevel(params_.max_concurrency);
+    std::this_thread::sleep_for(sleep_time);
+
+    CheckSequences();
+  }
+
+ private:
+  void CheckConcurrency()
+  {
+    if (params_.max_concurrency < 4) {
+      CHECK(stats_->num_active_infer_calls.load() == params_.max_concurrency);
```

Comment:
is load() needed because atomic variable? Does a bunch of old code need to be updated to use this?

Reply:
It's actually not needed here as its value gets loaded automatically via [std::atomic<T>::operator T](https://en.cppreference.com/w/cpp/atomic/atomic/operator_T). I think [std::atomic<T>::load](https://en.cppreference.com/w/cpp/atomic/atomic/load) is needed when printing and I incorrectly assumed it was needed here in the comparison operator.